src/backend/utils/adt/numeric.c

```diff
@@ -4535,7 +4536,15 @@ float4_numeric(PG_FUNCTION_ARGS)
 			PG_RETURN_NUMERIC(make_result(&const_pinf));
 	}
 
-	snprintf(buf, sizeof(buf), "%.*g", FLT_DIG, val);
+	/*
+	 * REAL datatype can store upto 4B of floating point data;
+	 * anything larger than that gets truncated unless
+	 * buf has extra places to store those extra_float_digits
+	 */
+	if (sql_dialect == SQL_DIALECT_TSQL && extra_float_digits > 0)
+		float_to_shortest_decimal_buf(val, buf);
+	else
+		snprintf(buf, sizeof(buf), "%.*g", FLT_DIG, val);
 
```

Comment:
similar logic can be found in float4out(); however in float4_numeric() the same was missing.

Reply:
we need more test cases to validate that this right check. This change implies that when dialect is TSQL, always try to use extra digits for REAL datatype when casting from float4 to numeric. Is it really the case for all possible casting scenario from float4 to numeric?